gk/main.c

```diff
@@ -918,12 +944,26 @@ process_pkts_front(uint8_t port_front, uint8_t port_back,
 				 * Notice that one needs to update
 				 * the Ethernet header.
 				 *
-				 * TODO Add sequential lock to deal with the
+				 * Use sequential lock to deal with the
 				 * concurrency between GK and LLS on the cached
 				 * Ethernet header.
 				 */
+
+				unsigned seq;
+				bool stale = false;
+
 				eth_cache = fib->u.gateway.eth_cache;
-				if (eth_cache == NULL || eth_cache->stale) {
+				if (eth_cache == NULL) {
+					drop_packet(pkt);
+					continue;
+				}
+
+				do {
+					seq = read_seqbegin(&eth_cache->lock);
+					stale = eth_cache->stale;
```

Comment:
1. BUG: I can't comment on that reads `eth_cache->eth_hdr`, but combine that line with the check of stale.

2. This issue also applies to the case `GK_FWD_NEIGHBOR_BACK_NET`, `GK_FWD_GATEWAY_FRONT_NET`, and `GK_FWD_NEIGHBOR_FRONT_NET`.

3. There's a lot of repeated code. Can't we avoid this with a function?